src/parser.rs

```diff
@@ -59,168 +98,231 @@ fn set_details<'a>(
     err
 }
 
-pub(crate) fn document(input: &str) -> IResult<&str, KdlDocument, KdlParseError<&str>> {
-    let (input, nodes) = many0(node)(input)?;
-    let (input, trailing) = all_whitespace(input)?;
-    let mut doc = KdlDocument::new();
-    doc.set_leading("");
-    doc.set_trailing(trailing);
-    *doc.nodes_mut() = nodes;
-    Ok((input, doc))
-}
-
-pub(crate) fn node(input: &str) -> IResult<&str, KdlNode, KdlParseError<&str>> {
-    let (input, leading) = all_whitespace(input)?;
-    let start = input;
-    let (input, ty) = opt(context("valid node type annotation", annotation))(input)?;
-    let (input, name) = context("valid node name", identifier)(input)?;
-    let (input, entries) = many0(context("valid node entry", entry))(input)?;
-    let (input, children) = opt(context("valid node children block", children))(input)?;
-    let (input, trailing) = context(
-        "valid node terminator",
-        cut(recognize(preceded(
-            many0(node_space),
-            alt((
-                terminated(recognize(tag(";")), opt(alt((linespace, eof)))),
-                alt((newline, single_line_comment, eof)),
-            )),
-        ))),
-    )(input)
-    .map_err(|e| {
-        set_details(
-            e,
-            start,
-            Some("parsed node"),
-            Some("Nodes can only be terminated by `;` or a valid line ending."),
-        )
-    })?;
-    let mut node = KdlNode::new(name);
-    node.set_leading(leading);
-    node.set_trailing(trailing);
-    node.ty = ty;
-    let ents = node.entries_mut();
-    *ents = entries;
-    if let Some((before, children)) = children {
-        let childs = node.children_mut();
-        *childs = Some(children);
-        node.set_before_children(before);
+pub(crate) fn document<'a: 'b, 'b>(
+    kdl_parser: &'b KdlParser<'a>,
+) -> impl Fn(&'a str) -> IResult<&'a str, KdlDocument, KdlParseError<&'a str>> + 'b {
```

Comment:
a lot of the signature hell here is from me trying to keep this "general" and not over-fit the signatures to what's currently needed by KdlParser.

Strictly speaking we could make the current KdlParser Copy and even 'static by just making it store the usize address of the input, which is the only thing we *really* need. This would eliminate the need for the 'b and even make it possible to have `fn parse` construct it for you. However if you ever want to put something non-Copy in there (or Really want the value of the full input) then you're going to have Problems.

One thing I didn't properly explore was whether the Fns could be made FnMuts, and KdlParser passed `&mut` -- this would allow the closures to mutate the KdlParsers, but I would say that mutating some global object randomly during parse is........ inadvisable.

Reply:
> and even make it possible to have fn parse construct it for you

(not actually 100% sure about this now that I think about it...)